### PR TITLE
MGXS API Changes for Arbitrary Reaction Channels and for Accurate Chi Computation

### PR DESCRIPTION
**Pull request by nelsonag**
This pull request adds two features to the MGXS Python API:

1. Adds ArbitraryXS and ArbitraryMatrixXS classes so the user can request MGXS vectors or energy transfer matrices for any reaction type that we can currently tally. For example, the user can request a (n,3n) MGXS or even an (n,Xt) MGXS. Note that the matrix MGXS are limited to reaction types that have at least one neutron as a product. The code will check for that. Why is this extra code worth its weight? Well, I'm not going to claim its crucial, but, it allows us to use the MGXS API to create MGXS for group-wise depletion libraries and it allows us to gather information we need to both fully fill out an ISOTXS CCCC file format while actually preserving neutrons (since ISOTXS only knows about (n,2n) reactions). Since ISOTXS is still decently common, and is the library format for DIF3D, this is already useful.

2. Some multigroup solvers perform an approximate weighting of isotopic Chi values to create a 'macroscopic' Chi. Such solvers include DIF3D and, everyone's favorite, the OpenMC MG solver (#1694). This approximate method weights the isotopic chi distributions by the isotopic macroscopic 1-group nu-fission cross section but it obtains the 1-group xs from the G-group data without flux-weighting (!!). This is done so that only a macroscopic Chi is needed during transport saving memory, code complexity, etc. This is generally acceptable for reactors that have your fission source dominated by one isotope. However, as is common in fast reactors, as you increase the # of important fissile isotopes this approximation begins to induce error (in our tests we found ~80pcm error due to this in a fast reactor design). This PR therefore also adds the ability to workaround such Chi normalization techniques. Specifically, the user now has access to an `apply_domain_chi` flag in the `Library.get_xsdata(...)`, `Library.create_mg_library(...)`, or `Library.create_mg_mode(...)` methods. If this flag is set to True, then every nuclide in a domain will actually receive the same Chi vector. This Chi vector is fission-source weighted across all isotopes, as it rigorously should be, and is the exact same as would be obtained when generating macroscopic MGXS from OpenMC. This The apply_domain_chi parameter defaults to False, meaning the MGXS API acts the same as it always has by providing nuclidic Chi values. Also note that this flag does nothing for macroscopic MGXS.

I have modified the unit tests for the new abstract XS types and manually verified that the x/s we get sum to what they should for scatter. I also repeated this for the matrix. This is why the results_true.dat for mgxs_library_no_nuclides includes a print of the reaction channel now. I left it there because I think its still valuable to have so we know the basis for the numbers we are looking at. I did not verify the nuclidic version of this to the same degree, but since that nuclidic behavior is inherited from upstream classes, I am not worried.

The `apply_domain_chi` flag is not in the regression tests. To actually test it would require an additional test; that additional burden on Travis didnt seem worth it for a single parameter that will be relatively niche. However, we should def consider this for inclusion when doing a full unit test suite in the future. Note, however, that I am currently using this feature now and can confirm it exactly reproduces the correct macroscopic Chi in downstream codes (both OpenMC MG and DIF3D).

Other changes include:
- Corrected the C++ representation of MT 180 ((n,3n2a)), as it was represented erroneously as (n,3n3a). I only found this because of the abstract xs unit tests discussed above.
- In reaction.py, the (n,n#) started at (n,n0) instead of (n,n1). Fixed.
- Cleaned up docstrings for MGXS Classes and Library that provided the acceptable values of the mgxs_type parameter. That list was becoming too long, even before the abstract XS types, and was limited to one line. I replaced it with a description.

### CODE REVIEW

**Comment by nelsonag** [comment]
lol, the +46,310 lines are due to the results_true.dat file. I guess I could convert that to a hashed result if anyone cares. However, my default is always to leave these files as plain text until you push me on it as I like having the least amount of work inbetween me and what I want to know :-)

**Comment by nelsonag** [comment]
The OMP=y, MPI=y, PHDF5=y, DAGMC=y build configurations is hitting the 50 minute time limit on Travis and failing. It failed after a restart too and so looks to be real. 

I dont need to test all the reaction types in the ArbitraryXS classes since all go through the same block of code for tallying, but, 1) this is probably the only place where we have a tally for every possible score it seems (based on the (n,3n2a) issue at least) and 2) this isn't really a runtime driver for the tests as those additional scores dont take _that_ much more time.

Is there something else we could clean up to buy more time?

**Comment by nelsonag** [comment]
Note I added one more method to this PR: MGXS.get_flux(...). I just found it helpful in finding what the weighting flux was for doing post-processing that wasn't worth main-lining in to the rest of the MGXS classes.

**Comment by nelsonag** [comment]
I just reduced the number of tested MGXS types in mgxs_library_no_nuclides and mgxs_library_nuclides in an effort to test the intended Python code but not over-do Travis' effort.

Now I'm doing it for most of those with non-zero scores in these tests, but for the level reactions i'm not including every level, just one.

**Comment by paulromano** [comment]
@nelsonag Sorry I haven't had a chance to look at this yet. Should have a review for you by tomorrow.

File: openmc/mgxs/mgxs.py [hunk 1]
```diff
@@ -3850,6 +3995,265 @@ def nu(self, nu):
...
+       \langle \sigma_{X,g'\rightarrow g} \phi \rangle &= \int_{r \in V} dr
+       \int_{4\pi} d\Omega' \int_{E_{g'}}^{E_{g'-1}} dE' \int_{E_g}^{E_{g-1}} dE
+       \; \chi(E) \sigma_X (r, E') \psi(r, E', \Omega')\\
+       \langle \phi \rangle &= \int_{r \in V} dr \int_{4\pi} d\Omega
+       \int_{E_g}^{E_{g-1}} dE \; \psi (r, E, \Omega) \\
+       \sigma_{X,g'\rightarrow g} &= \frac{\langle \sigma_{X,g'\rightarrow
+       g} \phi \rangle}{\langle \phi \rangle}
+       \end{aligned}
+
+    where :math:`_\sigma_X` is the requested reaction type of interest.
```
Comment (paulromano) [inline openmc/mgxs/mgxs.py:4155]:
```suggestion
    where :math:`\sigma_X` is the requested reaction type of interest.
```

File: openmc/mgxs/mgxs.py [hunk 2]
```diff
@@ -3850,6 +3995,265 @@ def nu(self, nu):
...
+    For a spatial domain :math:`V` and energy group :math:`[E_g,E_{g-1}]`, the
+    requested cross section is calculated as:
+
+    .. math::
+
+       \frac{\int_{r \in V} dr \int_{4\pi} d\Omega \int_{E_g}^{E_{g-1}} dE \;
+       \sigma_X (r, E) \psi (r, E, \Omega)}{\int_{r \in V} dr \int_{4\pi}
+       d\Omega \int_{E_g}^{E_{g-1}} dE \; \psi (r, E, \Omega)}
+
+    where :math:`_\sigma_X` is the requested reaction type of interest.
```
Comment (paulromano) [inline openmc/mgxs/mgxs.py:4024]:
```suggestion
    where :math:`\sigma_X` is the requested reaction type of interest.
```

File: tests/regression_tests/mgxs_library_no_nuclides/test.py [hunk 3]
```diff
@@ -48,7 +56,7 @@ def _get_results(self, hash_output=False):
             for mgxs_type in self.mgxs_lib.mgxs_types:
                 mgxs = self.mgxs_lib.get_mgxs(domain, mgxs_type)
                 df = mgxs.get_pandas_dataframe()
-                outstr += df.to_string() + '\n'
+                outstr += mgxs_type + '\n' + df.to_string() + '\n'
```
Comment (paulromano) [inline tests/regression_tests/mgxs_library_no_nuclides/test.py:59]: :+1: This does make the results a lot easier to figure out

**Review by paulromano — CHANGES_REQUESTED** [review]
Nice improvement of our MGXS capabilities; thanks @nelsonag! Just a few minor comments:

File: openmc/mgxs/mgxs.py [hunk 4]
```diff
@@ -1139,6 +1170,120 @@ def get_xs(self, groups='all', subdomains='all', nuclides='all',
...
+        # Determine which flux to obtain
+        # Step through in order of usefulness
+        tally = None
+        for key in ['flux', 'flux (tracklength)', 'flux (analog)']:
+            if key in self.tally_keys:
+                tally = self.tallies[key]
+                break
+        if tally is None:
+            msg = "MGXS of Type {} do not have an explicit weighting flux!"
+            raise ValueError(msg.format(self.__name__))
```
[the remark this answers is not in the document]
Comment (nelsonag) [reply openmc/mgxs/mgxs.py:1247]: Good catch, thanks! I used one of those for the first time last week on another project too!

**Comment by nelsonag** [comment]
Thanks for the review @paulromano! I just committed the requested changes!

**Review by paulromano — APPROVED** [review]

**Pull request merged**